omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,907 @@
+
+    vector<IndexType> final_level_counts(max_size + 1, exec);
+    auto i = 0;
+    while (true) {
+        auto done = true;
+        for (auto tid = 0; tid < num_threads; ++tid) {
+            if (i < level_count_sizes[tid]) {
+                const auto count = level_counts[tid][i];
+                final_level_counts[i] += count;
+                done = false;
+            }
+        }
+        if (done) {
+            break;
+        }
+        ++i;
+    }
+
+    return final_level_counts;
+}
+
+
+/**
+ * Implements the two intermediate phases of urcm,
+ * counting the nodes per level and computing the level offsets,
+ * as a prefix sum.
+ */
+template <typename IndexType>
+vector<IndexType> compute_level_offsets(std::shared_ptr<const OmpExecutor> exec,
+                                        const IndexType *const levels,
+                                        size_type num_vertices,
+                                        uint8 *const previous_component)
+{
+    auto counts = count_levels(exec, levels, previous_component, num_vertices);
+    components::prefix_sum(exec, &counts[0], counts.size());
+    return counts;
+}
+
+
+/**
+ * Helper macros for the helper function below.
+ */
+#if defined(__GNUG__) && defined(__x86_64__)
+// GCC can only be convinced to use cmovs here by inline asm.
+#define GKO_COMPARATOR(x, y)                                       \
+    {                                                              \
+        IndexType tmp;                                             \
+        const auto degree_x = degrees[neighbours[x]];              \
+        const auto degree_y = degrees[neighbours[y]];              \
+        asm("mov %0, %2\n\t"                                       \
+            "cmp %4, %3\n\t"                                       \
+            "cmova %1, %0\n\t"                                     \
+            "cmova %2, %1\n\t"                                     \
+            : "+r"(neighbours[x]), "+r"(neighbours[y]), "=&r"(tmp) \
+            : "r"(degree_x), "r"(degree_y)                         \
+            : "cc");                                               \
+    }
+#else
+#define GKO_COMPARATOR(x, y)                                         \
+    {                                                                \
+        const auto degree_x = degrees[neighbours[x]];                \
+        const auto degree_y = degrees[neighbours[y]];                \
+        const auto should_swap = degree_x > degree_y;                \
+        const auto tmp = neighbours[x];                              \
+        neighbours[x] = should_swap ? neighbours[y] : neighbours[x]; \
+        neighbours[y] = should_swap ? tmp : neighbours[y];           \
+    }
+#endif
+
+
+/**
+ * Helper fucntion for quickly sorting small vectors by degree.
+ */
+template <typename IndexType>
+inline void sort_by_degree(IndexType *const neighbours, size_type count,
```

Comment:
Such a small-size sorting implementation might actually be useful for more situations. Maybe we should generalize and extract it?

Reply:
I will do some tests concerning compilation given lambdas and how many sizes to match explicitly.

Reply:
In case you want some input, you could check the end of [this blog article](https://blog.reverberate.org/2020/05/29/hoares-rebuttal-bubble-sorts-comeback.html) or [this thesis](http://algo2.iti.kit.edu/documents/Theses/2019%20-%20Bachelor-Thesis_Jasper_Marianczuk_Engineering%20Faster%20Sorters%20for%20Small%20Sets%20of%20Items.pdf), both on small-size sorting